src/server/coprocessor/executor/selection.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// FIXME(andelf): remove following later
+#![allow(dead_code)]
+
+use std::rc::Rc;
+use std::collections::HashSet;
+
+use tipb::executor::Selection;
+use tipb::schema::ColumnInfo;
+use tipb::expression::{Expr, ExprType};
+use util::xeval::{Evaluator, EvalContext};
+use util::codec::number::NumberDecoder;
+
+use super::{Row, Executor};
+use super::super::Result;
+use super::super::endpoint::inflate_with_col;
+
+/// Collect all mentioned column ids in expr tree
+pub struct ExprColumnRefVisitor {
+    pub column_ids: HashSet<i64>,
+}
+
+impl ExprColumnRefVisitor {
+    pub fn new() -> ExprColumnRefVisitor {
+        ExprColumnRefVisitor { column_ids: HashSet::new() }
+    }
+
+    pub fn visit_expr<'e>(&mut self, expr: &'e Expr) -> Result<()> {
+        if expr.get_tp() == ExprType::ColumnRef {
+            self.column_ids.insert(try!(expr.get_val().decode_i64()));
+        } else {
+            for sub_expr in expr.get_children() {
+                try!(self.visit_expr(sub_expr));
+            }
+        }
+        Ok(())
+    }
+}
+
+pub struct SelectionExecutor<'a> {
+    conditions: Vec<Expr>,
+    columns: Vec<ColumnInfo>,
+    ctx: Rc<EvalContext>,
+    src: Box<Executor + 'a>,
+}
+
+impl<'a> SelectionExecutor<'a> {
+    pub fn new(mut meta: Selection,
+               ctx: Rc<EvalContext>,
+               columns_info: &[ColumnInfo],
+               src: Box<Executor + 'a>)
+               -> Result<SelectionExecutor<'a>> {
+        let conditions = meta.take_conditions().into_vec();
+
+        let mut visitor = ExprColumnRefVisitor::new();
+        for cond in &conditions {
+            try!(visitor.visit_expr(cond));
+        }
+
+        // FIXME(andelf): assume all items in columns_info are unique.
```

Comment:
first, tidb can guarantee the sql validity, so we can be sure that every columns used in condition will be exists in the column_info vec param.
second, tidb will guarantee every col in coumn_info vec param is unique.